domain/usecase/loan/service.go

```diff
@@ -34,12 +34,15 @@ func (s *Service) Borrow(u *entity.User, b *entity.Book) error {
 	if b.Quantity <= 0 {
 		return domain.ErrNotEnoughBooks
 	}
-	for _, v := range u.Books {
-		if v == b.ID {
-			return domain.ErrBookAlreadyBorrowed
-		}
+
+	_, err = u.GetBook(b.ID)
+	if err == nil {
+		return domain.ErrBookAlreadyBorrowed
+	}
+	err = u.AddBook(b.ID)
```

Comment:
Just a suggestion. You can move to AddBook the responsibility of check if this book is in the "user's book list". With this, you can remove the GetBook call above.